### PR TITLE
Introduce hardware pattern matching based on ONNXmetatypes

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Make hardware patterns matching using ONNXmetatypes

### Reason for changes

To make more robust hardware pattern mapping

### Related tickets

No

### Tests

Update existed tests

### CODE REVIEW

File: nncf/common/graph/graph.py [hunk 1]
```diff
@@ -32,6 +31,13 @@
 LayerName = str
 
 
+def get_edge_boundaries(match: List[str], graph: nx.DiGraph):
```
Comment (vshampor) [inline nncf/common/graph/graph.py:34]: Docstring, return type hint

File: nncf/common/graph/graph.py [hunk 2]
```diff
@@ -141,12 +147,13 @@ class NNCFGraphPatternIO:
...
     Describes the inputs and outputs of a subgraph in NNCFGraph.
     """
+
     def __init__(self, input_edges: List[NNCFGraphEdge], output_edges: List[NNCFGraphEdge]):
         self.input_edges = input_edges
         self.output_edges = output_edges
 
 
-#pylint:disable=too-many-public-methods
+# pylint:disable=too-many-public-methods
```
Comment (vshampor) [inline nncf/common/graph/graph.py:156]: This change is not required

File: nncf/common/graph/graph.py [hunk 3]
```diff
@@ -498,13 +504,13 @@ def visualize_graph(self, path: str):
...
             A.layout('dot')
-            png_path = os.path.splitext(path)[0]+'.png'
+            png_path = os.path.splitext(path)[0] + '.png'
             A.draw(png_path)
         except ImportError:
             nncf_logger.warning('Graphviz is not installed - only the .dot model visualization format will be used. '
                                 'Install pygraphviz into your Python environment and graphviz system-wide to enable '
                                 'PNG rendering.')
-        except Exception: #pylint:disable=broad-except
+        except Exception:  # pylint:disable=broad-except
```
Comment (vshampor) [inline nncf/common/graph/graph.py:513]: Same here

**Review by vshampor — CHANGES_REQUESTED** [review]
Basically good, minor changes required.

File: nncf/common/graph/graph.py (hunk 3 above)
Comment (kshpv) [reply nncf/common/graph/graph.py]: Done

File: nncf/common/graph/graph.py (hunk 2 above)
Comment (kshpv) [reply nncf/common/graph/graph.py]: Done

File: nncf/common/graph/graph.py (hunk 1 above)
Comment (kshpv) [reply nncf/common/graph/graph.py]: Done

File: nncf/common/utils/backend.py [hunk 4]
```diff
@@ -37,12 +38,20 @@ def infer_backend_from_model(model) -> BackendType:
...
     if torch is not None and isinstance(model, torch.nn.Module):
         return BackendType.TORCH
 
     if tensorflow is not None and isinstance(model, tensorflow.Module):
         return BackendType.TENSORFLOW
 
+    if onnx is not None and isinstance(model, onnx.ModelProto):
+        return BackendType.ONNX
+
     raise RuntimeError('Could not infer the backend framework from the model type because '
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/utils/backend.py:55]: Done

File: nncf/common/graph/graph.py (hunk 1 above)
Comment (vshampor) [reply nncf/common/graph/graph.py]: Making this a nested function reduces readability. Move this to a private method of NNCFGraph instead.

Comment (kshpv) [reply nncf/common/graph/graph.py]: should I make it static in that case?

File: nncf/common/utils/backend.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/common/utils/backend.py:59]: Done

File: nncf/common/graph/graph.py (hunk 1 above)
Comment (kshpv) [reply nncf/common/graph/graph.py]: Done

**Review by vshampor — APPROVED** [review]

**Pull request merged**